### PR TITLE
Gather cacheability logic into CachePolicy to provide an extension point for custom cache logic.

### PR DESCRIPTION
**Pull request by capoferro**
In order for the CachePolicy interface to be useful as an extension point, we need to gather all the logic that it's supposed to control into an implementation of CachePolicy, which is CachePolicyImpl in this PR. This PR is a large reorganization effort more than new functionality.

Commit Message: Gather cacheability logic into CachePolicy to provide an extension point for custom cache logic.
Additional Description:
Risk Level: Medium. CacheFilter is not a critical filter, but this is a very large change to CacheFilter.
Testing: Unit, integration

### CODE REVIEW

File: api/envoy/extensions/filters/http/cache/v3/cache.proto [hunk 1]
```diff
@@ -53,6 +54,10 @@ message CacheConfig {
   // [#extension-category: envoy.filters.http.cache]
   google.protobuf.Any typed_config = 1 [(validate.rules).any = {required: true}];
 
+  // CachePolicy extension that takes care of cache key, cacheability, cache
+  // freshness computations.
+  google.protobuf.Any typed_cache_policy_config = 5 [(validate.rules).any = {required: true}];
```
Comment (htuch) [inline api/envoy/extensions/filters/http/cache/v3/cache.proto:59]: TypedExtensionConfig

File: api/envoy/extensions/filters/http/cache/v3/cache_policy.proto [hunk 2]
```diff
@@ -0,0 +1,17 @@
+syntax = "proto3";
```
Comment (htuch) [inline api/envoy/extensions/filters/http/cache/v3/cache_policy.proto:1]: This is not an HTTP filter config, rather the config for a cache policy example, it should belong at top-level, under something like `api/envoy/extensions/http/cache_policy`

File: api/envoy/extensions/filters/http/cache/v3/cache.proto (hunk 1 above)
Comment (htuch) [reply api/envoy/extensions/filters/http/cache/v3/cache.proto:59]: I have no idea why `typed_config` is an `Any`, that was a mistake.

File: api/envoy/extensions/filters/http/cache/v3/cache_policy.proto (hunk 2 above)
Comment (htuch) [reply api/envoy/extensions/filters/http/cache/v3/cache_policy.proto:1]: This belongs in `api/envoy/extensions/cache_policy/cache_policy/v3/config.proto` actually, to parallel the existing cache storage extension. I assume this is some simple nop cache policy? You should add more descriptive config proto comments.

**Comment by jmarantz** [comment]
can you merge main?

**Comment by phlax** [comment]
@capoferro i see that the proto_format job is not uploading the diff fix correctly - i have a fix for this which i will PR shortly

**Comment by phlax** [comment]
and the error message about "tools/extensions/extensions_check.py" should read "tools/extensions/extensions_schema.yaml"

**Comment by phlax** [comment]
fix for `proto_format` diff is here #21896

File: api/envoy/extensions/filters/http/cache/v3/cache.proto [hunk 3]
```diff
@@ -51,8 +52,14 @@ message CacheConfig {
 
   // Config specific to the cache storage implementation.
   // [#extension-category: envoy.http.cache]
+  // TODO (capoferro): This should be TypedExtensionConfig, not Any
```
Comment (phlax) [inline api/envoy/extensions/filters/http/cache/v3/cache.proto:55]:
```suggestion
  // [#comment:TODO (capoferro): This should be TypedExtensionConfig, not Any]
```

same elsewhere

File: api/envoy/extensions/cache/cache_policy/v3/BUILD [hunk 4]
```diff
@@ -0,0 +1,9 @@
+# DO NOT EDIT. This file is generated by tools/proto_format/proto_sync.py.
+
+load("@envoy_api//bazel:api_build_system.bzl", "api_proto_package")
+
+licenses(["notice"])  # Apache 2
+
+api_proto_package(
+    deps = ["@com_github_cncf_udpa//udpa/annotations:pkg"],
```
Comment (phlax) [inline api/envoy/extensions/cache/cache_policy/v3/BUILD:8]:
```suggestion
    deps = ["@com_github_cncf_udpa//xds/annotations/v3:pkg"],
```

**Comment by jmarantz** [comment]
Looks like format still needs a tweak?

/wait

**Pull request closed without merging**